consensus/consensus-types/src/common.rs

```diff
@@ -10,5 +12,118 @@ pub type Round = u64;
 /// Author refers to the author's account address
 pub type Author = AccountAddress;
 
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct TransactionSummary {
+    pub sender: AccountAddress,
+    pub sequence_number: u64,
+}
+
+impl fmt::Display for TransactionSummary {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}:{}", self.sender, self.sequence_number,)
+    }
+}
+
 /// The payload in block.
-pub type Payload = Vec<SignedTransaction>;
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
+pub enum Payload {
+    DirectMempool(Vec<SignedTransaction>),
+    InQuorumStore(Vec<ProofOfStore>),
+}
+
+impl Payload {
+    pub fn new_empty() -> Self {
+        Payload::DirectMempool(Vec::new())
+    }
+
+    pub fn len(&self) -> usize {
+        match self {
+            Payload::DirectMempool(txns) => txns.len(),
+            Payload::InQuorumStore(_poavs) => todo!(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        match self {
+            Payload::DirectMempool(txns) => txns.is_empty(),
+            Payload::InQuorumStore(_poavs) => todo!(),
+        }
+    }
+}
+
+// TODO: What I really want is an iterator that isn't necessarily a vector (e.g., read lazily from RocksDB). This doesn't seem like the way.
+impl IntoIterator for Payload {
+    type Item = SignedTransaction;
+    type IntoIter = std::vec::IntoIter<Self::Item>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        match self {
+            Payload::DirectMempool(txns) => txns.into_iter(),
+            Payload::InQuorumStore(_poavs) => todo!(),
+        }
+    }
+}
+
+impl fmt::Display for Payload {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Payload::DirectMempool(txns) => {
+                write!(f, "InMemory txns: {}", txns.len())
+            }
+            Payload::InQuorumStore(_poavs) => todo!(),
+        }
+    }
+}
+
+/// The payload to filter.
+#[derive(Deserialize, Serialize, Clone, Debug, PartialEq, Eq)]
+pub enum PayloadFilter {
+    DirectMempool(Vec<TransactionSummary>),
+    InQuorumStore(Vec<ProofOfStore>),
```

Comment:
I think we do not need all the ProofOfStore, Vec<Digest> should be enough.

Reply:
So the plan here is to throw away the duplicate proofs (for corresponding batches) the same way consensus threw away the dup txns based on sequence numbers?
then indeed I think this can be just a Vec of HashValues corresponding to digests.